main/core.js

```diff
@@ -39,86 +37,62 @@ async function setup (/** @type {Context} */ ctx) {
       ? await dialog.showSaveDialog(win, opts)
       : await dialog.showSaveDialog(opts)
     if (filePath) {
-      await fs.writeFile(filePath, await core.logs.get())
+      await fs.writeFile(filePath, logs.get())
     }
   }
   await maybeMigrateFiles()
-  subscribeWithRetry(ctx, core).catch(console.error)
-  await start(core)
-}
-
-async function subscribeWithRetry (
-  /** @type {Context} */ ctx,
-  /** @type {Core} */ core
-) {
-  while (true) {
-    const controller = new AbortController()
-    try {
-      await subscribe(ctx, core, controller.signal)
-    } catch (err) {
-      controller.abort()
-      console.error(err)
-      await timers.setTimeout(1000)
-    }
-  }
-}
-
-async function subscribe (
-  /** @type {Context} */ ctx,
-  /** @type {Core} */ core,
-  /** @type {AbortSignal} */ signal
-) {
-  await Promise.all([
-    (async () => {
-      for await (const metrics of core.metrics.follow({ signal })) {
-        ctx.setTotalJobsCompleted(metrics.totalJobsCompleted)
-      }
-    })(),
-    (async () => {
-      const it = core.activity.follow({ nLines: 0, signal })
-      for await (const activity of it) {
-        ctx.recordActivity(activity)
-        detectChangeInOnlineStatus(activity)
-      }
-    })()
-  ])
+  await start(ctx)
 }
 
 /**
- * @param {Activity} activity
+ * @param {Context} ctx
  */
-function detectChangeInOnlineStatus (activity) {
-  if (
-    activity.type === 'info' &&
-    activity.message.includes('Saturn Node is online')
-  ) {
-    online = true
-  } else if (
-    activity.message === 'Saturn Node started.' ||
-    activity.message.includes('was able to connect') ||
-    activity.message.includes('will try to connect')
-  ) {
-    online = false
-  }
-}
-
-/**
- * @param {Core} core
- */
-async function start (core) {
+async function start (ctx) {
   assert(wallet.getAddress(), 'Core requires FIL address')
   console.log('Starting Core...')
 
-  const childProcess = fork(corePath, {
+  const childProcess = fork(corePath, ['--json'], {
     env: {
       ...process.env,
       FIL_WALLET_ADDRESS: wallet.getAddress(),
       CACHE_ROOT: consts.CACHE_ROOT,
       STATE_ROOT: consts.STATE_ROOT,
       DEPLOYMENT_TYPE: 'station-desktop'
+    },
+    stdio: ['pipe', 'pipe', 'pipe', 'ipc']
+  })
+
+  assert(childProcess.stdout)
+  childProcess.stdout.setEncoding('utf8')
+  childProcess.stdout.on('data', chunk => {
+    logs.push(chunk)
+    for (const line of chunk.split('\n').filter(Boolean)) {
+      const event = JSON.parse(line)
```

Comment:
Is it safe to assume that inter-process stdio will always split the data chunks on new-line boundary? If not, then we may receive a single line split across two or more chunks.

I am proposing to pipe stdio streams to `split2` before we parse it as ND-JSON. See https://github.com/mcollina/split2

Reply:
good idea! https://github.com/filecoin-station/desktop/pull/731